### PR TITLE
[SYCL] Make cl::reqd_work_group_size applicable to lambdas

### PR DESCRIPTION
**Pull request by MrSidims**
Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

File: clang/test/SemaSYCL/reqd-work-group-size.cpp [hunk 1]
```diff
@@ -96,17 +96,9 @@ void bar() {
...
 
-  // Support for reqd_work_group_size (and other SYCL attributes) that apply to
-  // lambda expressions is not implemented in clang yet.
-  // When it lands, the following code is expected to compile successfully.
-  //
-  // expected-error@+1 {{'reqd_work_group_size' attribute cannot be applied to types}}
-  kernel<class kernel_name9>([]() [[cl::reqd_work_group_size(32, 32, 32)]] {
-    f32x32x32();
-  });
   // While this case is not going to work (wrong syntax):
```
Comment (AGindinson) [inline clang/test/SemaSYCL/reqd-work-group-size.cpp]: Minor:
```suggestion
   // Wrong syntax:
```

File: clang/test/SemaSYCL/reqd-work-group-size.cpp [hunk 2]
```diff
@@ -96,17 +96,8 @@ void bar() {
     f32x32x1();
   });
 
-  // Support for reqd_work_group_size (and other SYCL attributes) that apply to
-  // lambda expressions is not implemented in clang yet.
-  // When it lands, the following code is expected to compile successfully.
-  //
-  // expected-error@+1 {{'reqd_work_group_size' attribute cannot be applied to types}}
-  kernel<class kernel_name9>([]() [[cl::reqd_work_group_size(32, 32, 32)]] {
```
Comment (asavonic) [inline clang/test/SemaSYCL/reqd-work-group-size.cpp]: Maybe leave this as a positive test case?

Comment (MrSidims) [reply clang/test/SemaSYCL/reqd-work-group-size.cpp]: I believe codegen test case added is enough.

Comment (asavonic) [reply clang/test/SemaSYCL/reqd-work-group-size.cpp]: I suggest to either keep it as a positive test, or remove all positive tests from the file.

Comment (MrSidims) [reply clang/test/SemaSYCL/reqd-work-group-size.cpp]: Done

File: clang/include/clang/Basic/AttributeCommonInfo.h [hunk 3]
```diff
@@ -152,7 +152,14 @@ class AttributeCommonInfo {
     // FIXME: Eventually we want to do a list here populated via tablegen.  But
     // we want C++ attributes to be permissible on Lambdas, and get propagated
     // to the call operator declaration.
-    return getParsedKind() == AT_SYCLIntelKernelArgsRestrict;
+    // FIXME: Attribute lambda capturing is not a part of SYCL 1.2.1. We shall
```
Comment (asavonic) [inline clang/include/clang/Basic/AttributeCommonInfo.h]: What do you mean by "is not a part of SYCL 1.2.1"? 
The spec is vague and doesn't explicitly mention how the attributes should be used:

> In SYCL, these attributes are legal on device functions and their specification is propagated down to any caller of those device functions

If a lambda function is considered to be a device function, then I don't see any problem with the spec.

Comment (MrSidims) [reply clang/include/clang/Basic/AttributeCommonInfo.h]: Ok, I agree.

**Pull request merged**